contributors/design-proposals/pod-resource-management.md

```diff
@@ -1,114 +1,480 @@
+will attempt to limit the abilty for a pod in a lower QoS tier to burst utilization
+of a compressible resource that was requested by a pod in a higher QoS tier.
+
+The `kubelet` will support a flag `experimental-qos-reserved` that
+takes a set of percentages per incompressible resource that controls how the
+QoS cgroup sandbox attempts to reserve resources for its tier.  It attempts
+to reserve requested resources to exclude pods from lower OoS classes from
+using resources requested by higher QoS classes. The flag will accept values
+in a range from 0-100%, where a value of `0%` instructs the `kubelet` to attempt
+no reservation, and a value of `100%` will instruct the `kubelet` to attempt to
+reserve the sum of requested resource across all pods on the node.  The `kubelet`
+initially will only support `memory`.  The default value per incompressible
+resource if not specified is for no reservation to occur for the incompressible
+resource.
+
+Prior to starting a pod, the `kubelet` will attempt to update the
+QoS cgroup sandbox associated with the lower QoS tier(s) in order
+to prevent consumption of the requested resource by the new pod.
+For example, prior to starting a Guaranteed pod, the Burstable
+and BestEffort QoS cgroup sandboxes are adjusted.  For resource
+specific details, and concerns, see the sections per resource that
+follow.
+
+The `kubelet` will allocate resources to the QoS level cgroup
+dynamically in response to the following events:
+
+1. kubelet startup/recovery
+1. prior to creation of the pod level cgroup
+1. after deletion of the pod level cgroup
+1. at periodic intervals to reach `experimental-qos-reserved`
+heurisitc that converge to a desired state.
+
+All writes to the QoS level cgroup sandboxes are protected via a
+common lock in the kubelet to ensure we do not have multiple concurrent
+writers to this tier in the hierarchy.
+
+### QoS level CPU allocation
+
+The `BestEffort` cgroup sandbox is statically configured as follows:
+
+```
+ROOT/BestEffort/cpu.shares = 2
+```
+
+This ensures that allocation of CPU time to pods in this QoS class
+is given the lowest priority.
+
+The `Burstable` cgroup sandbox CPU share allocation is dynamic based
+on the set of pods currently scheduled to the node.  
+
+```
+ROOT/Burstable/cpu.shares = max(sum(Burstable pods cpu requests, 2)
+```
+
+The Burstable cgroup sandbox is updated dynamically in the exit
+points described in the previous section.  Given the compressible
+nature of CPU, and the fact that cpu.shares are evaluated via relative
+priority, the risk of an update being incorrect is minimized as the `kubelet`
+converges to a desired state.  Failure to set `cpu.shares` at the QoS level
+cgroup would result in `500m` of cpu for a Guaranteed pod to have different
+meaning than `500m` of cpu for a Burstable pod in the current hierarchy.  This
+is because the default `cpu.shares` value if unspecified is `1024` and `cpu.shares`
```

Comment:
We can avoid this issue by defaulting new cgroups to a much lower value like `100m`.